src/framebuffer/swtfb_client.rs

```diff
@@ -0,0 +1,224 @@
+//! This implements the IPC part of a rM2Framebuffer client to interact with this server:
+//! https://github.com/ddvk/remarkable2-framebuffer
+//!
+//! The client is developed according to the spec here:
+//! https://github.com/ddvk/remarkable2-framebuffer/issues/11
+
+use super::mxcfb::mxcfb_update_data;
+use crate::framebuffer::screeninfo::{FixScreeninfo, VarScreeninfo};
+use memmap2::{MmapOptions, MmapRaw};
+use std::ffi::{c_void, CString};
+use std::fs::{File, OpenOptions};
+use std::io::Error as IoError;
+use std::os::unix::prelude::AsRawFd;
+use std::{env, mem, ptr};
+
+const SWTFB_MESSAGE_QUEUE_ID: i32 = 0x2257c;
+
+pub const WIDTH: i32 = 1404;
+pub const HEIGHT: i32 = 1872;
+
+#[allow(non_upper_case_globals)]
+pub const maxWidth: i32 = 1404;
+#[allow(non_upper_case_globals)]
+pub const maxHeight: i32 = 1872;
+pub const BUF_SIZE: i32 = maxWidth * maxHeight * std::mem::size_of::<u16>() as i32; // hardcoded size of display mem for rM2
+const SEM_WAIT_TIMEOUT: i32 = 200000000; /* 200 * 1000 * 1000, e.g. 200ms */
+
+/// long on 32 bit is 4 bytes as well!!
+#[derive(Debug, Clone, Copy)]
+#[repr(i32)]
+#[allow(non_camel_case_types)]
+#[allow(dead_code)]
+pub enum MSG_TYPE {
+    INIT_t = 1,
+    UPDATE_t = 2,
+    XO_t = 3,
+    WAIT_t = 4,
+}
+
+#[derive(Debug, Clone, Copy)]
+#[allow(non_camel_case_types, dead_code)]
+pub struct xochitl_data {
+    pub x1: i32,
+    pub y1: i32,
+    pub x2: i32,
+    pub y2: i32,
+
+    pub waveform: i32,
+    pub flags: i32,
+}
+
+#[derive(Debug, Clone, Copy)]
+#[allow(non_camel_case_types, dead_code)]
+pub struct wait_sem_data {
+    /// C string
+    pub sem_name: [u8; 512],
+}
+
+/// MSG_TYPE has to match swtfb_update_data !!!
+#[derive(Clone, Copy)]
+#[allow(non_camel_case_types, dead_code)]
+pub struct swtfb_update {
```

Comment:
Should these be `#[repr(C)]`?

Reply:
Wow! I somehow thought I thought of this, but It seems I never added it (or removed it during debugging?!?). Great catch. Making it explicit is always a lot better. Didn't even know that structs in rust have the memory layout as in C by default (maybe by chance)?

Reply:
Should be fixed with a0c0ea9.